tests/test_otel.py

```diff
@@ -0,0 +1,367 @@
+            add_header "X-Otel-Tracestate" $http_tracestate;
+            return 204;
+        }
+
+        location /404 {
+            return 404;
+        }
+    }
+}
+
+"""
+
+(trace_id, span_id) = ("0af7651916cd43dd8448eb211c80319c", "b9c7c989f97918e1")
+context = {
+    "Traceparent": f"00-{trace_id}-{span_id}-01",
+    "Tracestate": "congo=ucfJifl5GOE,rojo=00f067aa0ba902b7",
+}
+
+
+def decode_id(span, value):
+    if value in ["trace_id", "span_id"]:
+        return binascii.hexlify(getattr(span, value)).decode("utf-8")
+    return value
+
+
+def get_attr(span, attr, atype):
+    for value in (a.value for a in span.attributes if a.key == attr):
+        return getattr(value, atype)
+
+
+def get_http09(scheme, port, path, logger):
+    def do_get(sock, path):
+        http_send = f"GET {path}\n".encode()
+        logger.debug(f"{http_send=}")
+        sock.sendall(http_send)
+        http_recv = sock.recv(1024)
+        logger.debug(f"{http_recv=}")
+        return http_recv.decode("utf-8")
+
+    with socket.create_connection(("127.0.0.1", port)) as sock:
+        if scheme == "https":
+            ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+            ctx.check_hostname = False
+            ctx.verify_mode = ssl.CERT_NONE
+            with ctx.wrap_socket(sock, server_hostname="127.0.0.1") as ssock:
+                recv = do_get(ssock, path)
+        else:
+            recv = do_get(sock, path)
+    return recv
+
+
+def get_response(logger, http_ver, scheme, path, headers):
+    port = 18443 if scheme == "https" else 18080
+    if http_ver == 0:
+        return get_http09(scheme, port, path, logger)
+    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    with niquests.Session(multiplexed=True) as s:
+        if http_ver == 3:
+            assert scheme == "https", "Only https:// URLs are supported."
+            s.quic_cache_layer.add_domain("127.0.0.1", port)
+        resp = s.get(
+            f"{scheme}://127.0.0.1:{port}{path}", headers=headers, verify=False
+        )
+    return resp
+
+
+@pytest.fixture
+def batches(trace_service):
+    for _ in range(20):
+        if len(trace_service.spans):
+            break
+        time.sleep(0.1)
+    else:
+        assert len(trace_service.spans) > 0, "No spans received"
+    yield trace_service.spans
+    trace_service.spans.clear()
+
+
+@pytest.fixture
+def span(batches):
+    assert len(batches) == 1, "Expect 1 batch"
+    assert len(batches[0][0].scope_spans[0].spans) == 1, "Expect 1 span"
+    return batches[0][0].scope_spans[0].spans[0]
+
+
+@pytest.fixture
+def response(logger, http_ver, scheme, path, headers):
+    return get_response(logger, http_ver, scheme, path, headers)
+
+
+@pytest.fixture
+def nresponses(logger, http_ver, scheme, path, headers, n):
+    out = []
+    for _ in range(n):
+        out.append(get_response(logger, http_ver, scheme, path, headers))
+    return out
+
+
+@pytest.mark.usefixtures("trace_service", "otelcol", "nginx")
```

Comment:
"otelcol" should be launched automatically, whenever "trace_service" is used, i.e. the code in this file doesn't need to know about "otelcol".